xarray/core/dataset.py

```diff
@@ -6857,146 +6856,124 @@ def polyfit(
         RankWarning
             The rank of the coefficient matrix in the least-squares fit is deficient.
-            The warning is not raised with in-memory (not dask) data and `full=True`.
+            The warning is not raised with `full=True`.
 
         See Also
         --------
         numpy.polyfit
         numpy.polyval
         xarray.polyval
         """
-        variables = {}
-        skipna_da = skipna
-
-        x = get_clean_interp_index(self, dim, strict=False)
-        xname = "{}_".format(self[dim].name)
-        order = int(deg) + 1
-        lhs = np.vander(x, order)
-
-        if rcond is None:
-            rcond = (
-                x.shape[0] * np.core.finfo(x.dtype).eps  # type: ignore[attr-defined]
-            )
-
-        # Weights:
         if w is not None:
             if isinstance(w, Hashable):
                 w = self.coords[w]
             w = np.asarray(w)
             if w.ndim != 1:
                 raise TypeError("Expected a 1-d array for weights.")
-            if w.shape[0] != lhs.shape[0]:
-                raise TypeError("Expected w and {} to have the same length".format(dim))
-            lhs *= w[:, np.newaxis]
+            if w.shape[0] != self[dim].shape[0]:
+                raise TypeError(f"Expected w and {dim} to have the same length")
 
-        # Scaling
-        scale = np.sqrt((lhs * lhs).sum(axis=0))
-        lhs /= scale
-
-        degree_dim = utils.get_temp_dimname(self.dims, "degree")
+        order = int(deg) + 1
+        degrees = np.arange(deg, -1, -1)
 
-        rank = np.linalg.matrix_rank(lhs)
+        x = get_clean_interp_index(self, dim, strict=False)
 
+        # sizes and names for varying polyfit outputs
+        output_sizes = {"degree": order}
         if full:
-            rank = xr.DataArray(rank, name=xname + "matrix_rank")
-            variables[rank.name] = rank
-            sing = np.linalg.svd(lhs, compute_uv=False)
-            sing = xr.DataArray(
-                sing,
-                dims=(degree_dim,),
-                coords={degree_dim: np.arange(rank - 1, -1, -1)},
-                name=xname + "singular_values",
-            )
-            variables[sing.name] = sing
+            output_core_dims = [("degree",), (), (), ("degree",), ()]
+            output_vars = [
+                "{name}polyfit_coefficients",
+                "{name}polyfit_residuals",
+                f"{dim}_matrix_rank",
+                f"{dim}_singular_values",
+                f"{dim}_rcond",
+            ]
+        elif cov and not full:
+            output_core_dims = [("degree",), ("cov_i", "cov_j")]
+            output_vars = [
+                "{name}polyfit_coefficients",
+                "{name}polyfit_covariance",
+            ]
+            output_sizes["cov_i"] = order
+            output_sizes["cov_j"] = order
+        else:
+            output_core_dims = [("degree",)]
+            output_vars = ["{name}polyfit_coefficients"]
+
+        def _wrapper(x, y):
```

Comment:
It might not be worth it, but you could avoid a `if` conditional in the inner loop as follows:

```python
def _wrapper_skipna(x, y):
    ...

def _wrapper_noskipna(x, y):
    ...

if skipna:
    _wrapper = _wrapper_skipna
else:
    _wrapper = _wrapper_noskipna

```